File: src/backends/backend.py
import abc

class Backend(abc.ABC):
    """Abstract base class for a storage backend."""

    @abc.abstractmethod
    def create_bucket(self, bucket_name, owner=None):
        """Create a bucket."""
        pass

    @abc.abstractmethod
    def delete_bucket(self, bucket_name):
        """Delete a bucket."""
        pass

    @abc.abstractmethod
    def bucket_exists(self, bucket_name):
        """Check if a bucket exists."""
        pass

    @abc.abstractmethod
    def get_bucket_owner(self, bucket_name):
        """Get the owner of a bucket."""
        pass

    @abc.abstractmethod
    def change_bucket_owner(self, bucket_name, new_owner):
        """Change the owner of a bucket."""
        pass

    @abc.abstractmethod
    def create_user(self, access_key, secret_key, role=None, user_id=None, group_id=None):
        """Create a user."""
        pass

    @abc.abstractmethod
    def delete_user(self, access_key):
        """Delete a user."""
        pass

    @abc.abstractmethod
    def update_user(self, access_key, secret_key=None, user_id=None, group_id=None):
        """Update a user."""
        pass

    @abc.abstractmethod
    def user_exists(self, access_key):
        """Check if a user exists."""
        pass

File: pkg/backends/backend.go
package backends

import "context"

// Backend defines the interface for S3-compatible storage backends
type Backend interface {
	// TestConnection tests the connection to the backend
	TestConnection(ctx context.Context) error

	// Bucket operations
	CreateBucket(ctx context.Context, bucketName string, owner *string) error
	DeleteBucket(ctx context.Context, bucketName string) error
	BucketExists(ctx context.Context, bucketName string) (bool, error)
	GetBucketOwner(ctx context.Context, bucketName string) (string, error)
	ChangeBucketOwner(ctx context.Context, bucketName, newOwner string) error

	// User operations
	CreateUser(ctx context.Context, accessKey, secretKey string, role *string, userID, groupID *int) error
	DeleteUser(ctx context.Context, accessKey string) error
	UpdateUser(ctx context.Context, accessKey string, secretKey *string, userID, groupID *int) error
	UserExists(ctx context.Context, accessKey string) (bool, error)

	// GetEndpointURL returns the configured endpoint URL
	GetEndpointURL() string
}

// Config holds common backend configuration
type Config struct {
	EndpointURL string
	AccessKey   string
	SecretKey   string
}

// NewBackend creates a new backend instance based on the backend name
func NewBackend(name string, config Config) (Backend, error) {
	switch name {
	case "versitygw":
		return NewVersityGW(config), nil
	case "minio":
		return NewMinIO(config), nil
	case "garage":
		return NewGarage(config), nil
	default:
		return nil, ErrUnsupportedBackend{Backend: name}
	}
}

// ErrUnsupportedBackend is returned when an unknown backend is requested
type ErrUnsupportedBackend struct {
	Backend string
}

func (e ErrUnsupportedBackend) Error() string {
	return "unsupported backend: " + e.Backend
}
